File: typescript/tests/test_keys.ts
import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";
import { Azex } from "../src/client.js";
import type { APIKey, APIKeyList } from "../src/types/keys.js";

const mockKey: APIKey = {
  uid: "key_abc123",
  name: "prod-key",
  prefix: "sk_live_",
  status: "active",
  rpm_limit: 100,
  created_at: "2026-03-26T00:00:00Z",
  last_used_at: null,
  key: "sk_live_abc123_secret",
};

const mockKeyList: APIKeyList = {
  items: [
    { uid: "key_abc123", name: "prod-key", prefix: "sk_live_", status: "active", created_at: "2026-03-26T00:00:00Z" },
    { uid: "key_def456", name: "test-key", prefix: "sk_live_", status: "suspended", created_at: "2026-03-25T00:00:00Z" },
  ],
  total: 2,
};

describe("keys resource", () => {
  let client: Azex;
  let fetchSpy: ReturnType<typeof vi.spyOn>;

  beforeEach(() => {
    client = new Azex({ apiKey: "sk_test_1234", maxRetries: 0 });
    fetchSpy = vi.spyOn(globalThis, "fetch");
  });

  afterEach(() => {
    vi.restoreAllMocks();
  });

  it("create sends POST to /api/v1/keys with params", async () => {
    fetchSpy.mockResolvedValueOnce(
      new Response(JSON.stringify(mockKey), { status: 200 }),
    );

    const result = await client.keys.create({ name: "prod-key", rpm_limit: 100 });

    expect(result.uid).toBe("key_abc123");
    expect(result.key).toBe("sk_live_abc123_secret");

    const [url, init] = fetchSpy.mock.calls[0] as [string, RequestInit];
    expect(url).toContain("/api/v1/keys");
    expect(init.method).toBe("POST");
    const body = JSON.parse(init.body as string);
    expect(body.name).toBe("prod-key");
    expect(body.rpm_limit).toBe(100);
  });

  it("list sends GET to /api/v1/keys", async () => {
    fetchSpy.mockResolvedValueOnce(
      new Response(JSON.stringify(mockKeyList), { status: 200 }),
    );

    const result = await client.keys.list();

    expect(result.total).toBe(2);
    expect(result.items).toHaveLength(2);
    expect(result.items[0].name).toBe("prod-key");
    expect(result.items[1].status).toBe("suspended");

    const [url, init] = fetchSpy.mock.calls[0] as [string, RequestInit];
    expect(url).toContain("/api/v1/keys");
    expect(init.method).toBe("GET");
  });

  it("revoke sends DELETE to /api/v1/keys/:uid", async () => {
    fetchSpy.mockResolvedValueOnce(
      new Response(null, { status: 204, headers: { "Content-Length": "0" } }),
    );

    await client.keys.revoke("key_abc123");

    const [url, init] = fetchSpy.mock.calls[0] as [string, RequestInit];
    expect(url).toContain("/api/v1/keys/key_abc123");
    expect(init.method).toBe("DELETE");
  });

  it("update sends PATCH to /api/v1/keys/:uid", async () => {
    const updated = { ...mockKey, name: "renamed-key", key: undefined };
    fetchSpy.mockResolvedValueOnce(
      new Response(JSON.stringify(updated), { status: 200 }),
    );

    const result = await client.keys.update("key_abc123", { name: "renamed-key" });

    expect(result.name).toBe("renamed-key");

    const [url, init] = fetchSpy.mock.calls[0] as [string, RequestInit];
    expect(url).toContain("/api/v1/keys/key_abc123");
    expect(init.method).toBe("PATCH");
    const body = JSON.parse(init.body as string);
    expect(body.name).toBe("renamed-key");
  });

  it("suspend sends POST to /api/v1/keys/:uid/suspend", async () => {
    const suspended = { ...mockKey, status: "suspended" as const, key: undefined };
    fetchSpy.mockResolvedValueOnce(
      new Response(JSON.stringify(suspended), { status: 200 }),
    );

    const result = await client.keys.suspend("key_abc123");

    expect(result.status).toBe("suspended");

    const [url, init] = fetchSpy.mock.calls[0] as [string, RequestInit];
    expect(url).toContain("/api/v1/keys/key_abc123/suspend");
    expect(init.method).toBe("POST");
  });

  it("resume sends POST to /api/v1/keys/:uid/resume", async () => {
    const resumed = { ...mockKey, status: "active" as const, key: undefined };
    fetchSpy.mockResolvedValueOnce(
      new Response(JSON.stringify(resumed), { status: 200 }),
    );

    const result = await client.keys.resume("key_abc123");

    expect(result.status).toBe("active");

    const [url, init] = fetchSpy.mock.calls[0] as [string, RequestInit];
    expect(url).toContain("/api/v1/keys/key_abc123/resume");
    expect(init.method).toBe("POST");
  });

  it("throws NotFoundError on 404", async () => {
    fetchSpy.mockResolvedValueOnce(
      new Response(JSON.stringify({ error: "Key not found" }), { status: 404 }),
    );

    await expect(client.keys.revoke("nonexistent")).rejects.toMatchObject({
      name: "NotFoundError",
      statusCode: 404,
    });
  });
});

File: python/tests/test_keys.py
"""Tests for API key management resource."""

from __future__ import annotations

import json

import pytest
import respx
import httpx

from azex import Azex
from azex._errors import NotFoundError
from azex.types.keys import APIKey, APIKeyList


KEY_FIXTURE = {
    "uid": "key_01J8XRTN3BEH4JBC8TZJX71XGG",
    "name": "prod-key",
    "prefix": "sk_live_abc",
    "status": "active",
    "rpm_limit": 100,
    "created_at": "2026-03-26T00:00:00Z",
    "last_used_at": None,
    "key": "sk_live_abc...secretpart",
}

KEY_FIXTURE_NO_KEY = {k: v for k, v in KEY_FIXTURE.items() if k != "key"}

LIST_FIXTURE = {
    "items": [KEY_FIXTURE_NO_KEY],
    "total": 1,
}


@respx.mock
def test_keys_create() -> None:
    respx.post("https://api.azex.ai/api/v1/keys").mock(
        return_value=httpx.Response(200, json=KEY_FIXTURE)
    )
    client = Azex(api_key="test-key")
    key = client.keys.create(name="prod-key", rpm_limit=100)
    assert isinstance(key, APIKey)
    assert key.uid == "key_01J8XRTN3BEH4JBC8TZJX71XGG"
    assert key.key == "sk_live_abc...secretpart"
    assert key.status == "active"
    assert key.rpm_limit == 100

    sent = json.loads(respx.calls.last.request.content)
    assert sent["name"] == "prod-key"
    assert sent["rpm_limit"] == 100


@respx.mock
def test_keys_list() -> None:
    respx.get("https://api.azex.ai/api/v1/keys").mock(
        return_value=httpx.Response(200, json=LIST_FIXTURE)
    )
    client = Azex(api_key="test-key")
    result = client.keys.list()
    assert isinstance(result, APIKeyList)
    assert result.total == 1
    assert len(result.items) == 1
    assert result.items[0].name == "prod-key"


@respx.mock
def test_keys_revoke() -> None:
    uid = "key_01J8XRTN3BEH4JBC8TZJX71XGG"
    respx.delete(f"https://api.azex.ai/api/v1/keys/{uid}").mock(
        return_value=httpx.Response(204)
    )
    client = Azex(api_key="test-key")
    result = client.keys.revoke(uid)
    assert result is None


@respx.mock
def test_keys_update() -> None:
    uid = "key_01J8XRTN3BEH4JBC8TZJX71XGG"
    updated = {**KEY_FIXTURE_NO_KEY, "name": "renamed-key"}
    respx.patch(f"https://api.azex.ai/api/v1/keys/{uid}").mock(
        return_value=httpx.Response(200, json=updated)
    )
    client = Azex(api_key="test-key")
    key = client.keys.update(uid, name="renamed-key")
    assert key.name == "renamed-key"

    sent = json.loads(respx.calls.last.request.content)
    assert sent["name"] == "renamed-key"


@respx.mock
def test_keys_suspend() -> None:
    uid = "key_01J8XRTN3BEH4JBC8TZJX71XGG"
    suspended = {**KEY_FIXTURE_NO_KEY, "status": "suspended"}
    respx.post(f"https://api.azex.ai/api/v1/keys/{uid}/suspend").mock(
        return_value=httpx.Response(200, json=suspended)
    )
    client = Azex(api_key="test-key")
    key = client.keys.suspend(uid)
    assert key.status == "suspended"


@respx.mock
def test_keys_resume() -> None:
    uid = "key_01J8XRTN3BEH4JBC8TZJX71XGG"
    active = {**KEY_FIXTURE_NO_KEY, "status": "active"}
    respx.post(f"https://api.azex.ai/api/v1/keys/{uid}/resume").mock(
        return_value=httpx.Response(200, json=active)
    )
    client = Azex(api_key="test-key")
    key = client.keys.resume(uid)
    assert key.status == "active"


@respx.mock
def test_keys_revoke_not_found() -> None:
    uid = "nonexistent"
    respx.delete(f"https://api.azex.ai/api/v1/keys/{uid}").mock(
        return_value=httpx.Response(404, json={"error": "key not found"})
    )
    client = Azex(api_key="test-key", max_retries=0)
    with pytest.raises(NotFoundError) as exc_info:
        client.keys.revoke(uid)
    assert exc_info.value.status_code == 404


@respx.mock
def test_keys_create_without_rpm_limit() -> None:
    no_rpm = {**KEY_FIXTURE, "rpm_limit": None}
    respx.post("https://api.azex.ai/api/v1/keys").mock(
        return_value=httpx.Response(200, json=no_rpm)
    )
    client = Azex(api_key="test-key")
    key = client.keys.create(name="dev-key")
    assert key.rpm_limit is None
    # rpm_limit should NOT be in sent body
    sent = json.loads(respx.calls.last.request.content)
    assert "rpm_limit" not in sent
